### ISSUE TITLE
userEvent.click triggers warning that it must be wrapped in act

### ISSUE THREAD

**Issue by shayantabrizi** [testing-library/user-event#716]
In [this](https://codesandbox.io/s/act-test-bzbdt?file=/src/index.test.js) sample code, after running tests I get "Warning: An update to Test inside a test was not wrapped in act(...)." I have seen the kentcdodds' blog and also other sources and as far as I know I have done everything right. Isn't this a bug of userEvent.click?

Code:
```
import * as React from 'react';
import {render} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import {useState} from "react";
import {Button} from '@material-ui/core';

function Test() {
    const [result, setResult] = useState(0);

    const handle = () => {
        new Promise(resolve => {
            resolve(1);
        }).then((result) => {
            setResult(1);
        });
    };

    return (
        <>
            <Button
                className="test-button"
                onClick={() => {
                    handle();
                }}
            />
            <div>{result}</div>
        </>
    );
}

test('Test', () => {
    render(<Test/>);
    userEvent.click(document.querySelector('.test-button'));
});
```

**Comment by ph-fritsche** [comment]
`user-event` dispatches the events per `fireEvent`.
When you import `@testing-library/react` those calls get automatically wrapped in `act` so a state change in your event handler is already wrapped.

But you trigger the state change asynchronously. Therefore it happens after the event handler and the `act` finished.
This is not a bug but to be expected.

**Comment by shayantabrizi** [comment]
So, could you give an advice on how to solve the problem?

**Comment by ph-fritsche** [comment]
Set state synchronously or `await` the change.
https://codesandbox.io/s/act-test-forked-2vsb2?file=/src/variantA.test.js

**Comment by shayantabrizi** [comment]
Thanks a lot. But, about variant B, isn't the warning misleading? Why should having a statement after the click have an effect on whether `an update is wrapped in act` or not?

**Comment by shayantabrizi** [comment]
Also, wrapping the userEvent in `await act` solves the problem. If `userEvent.click` is actually wrapped in an `act`, shouldn't wrapping again be a no action and have no effects?
```
  await act(async () => userEvent.click(screen.getByRole("button")));
```

**Comment by ph-fritsche** [comment]
If you searched the issues you would have found multiple closed issues with [explanations](https://github.com/testing-library/user-event/issues/128#issuecomment-720515712) on this topic:
> See https://developer.mozilla.org/en-US/docs/Web/JavaScript/EventLoop
When you use `setTimeout` or `Promise` you give up control over the order of operations. They will be added somewhere down the event loop and your code returns instantly.
If your code is simple, you can, of course, predict in which order the statements will be executed. But you should not rely on that.

**Comment by shayantabrizi** [comment]
Thanks a lot.

**Comment by babramczyk** [comment]
@ph-fritsche could you explain how `@testing-library/react` does this wrapping?

I’ve upgraded to Jest 27, and all latest versions of `@testing-library` deps. Now, some tests throw `act` warnings for something simple like `userEvent.click`

When I debug and step through, the final `fireEvent` that `user-event` fires goes straight to `@testing-library/dom`, and does **not** go through the RTL wrapper that uses `act`

Wrapping these specific events that result in the `act` warning fixes it.

**Comment by ph-fritsche** [comment]
The wrapping is done per configuration when importing the package.
https://github.com/testing-library/react-testing-library/blob/84851dc660d49543707ba9eead42395c45ec06e2/src/pure.js#L11-L26

**Comment by babramczyk** [comment]
@ph-fritsche thanks for pointing that out

I’m having some very weird troubles still.

I have a test ([here is a CodeSandbox](https://gitlab.com/wonder.me/wonder-web/-/merge_requests/632#fe29c668c360e8c36bfb871ab60ffffd596cebda) with the exact test). The test passed on CodeSandbox, but fails locally for the `works without act` test.

Do you have any thoughts here? I’ve tried isolating everything I can from my local (i.e. removing Jest setup files, using the exact deps in CodeSandbox as I have locally), but can’t seem to find the discrepancy, and why a manual `act` is necessary locally.

**Comment by ph-fritsche** [comment]
> I have a test ([here is a CodeSandbox](https://gitlab.com/wonder.me/wonder-web/-/merge_requests/632#fe29c668c360e8c36bfb871ab60ffffd596cebda) with the exact test). The test passed on CodeSandbox, but fails locally for the `works without act` test.

The link is points to a PR on Gitlab and is dead (404) - either the resource has been removed or it is hidden due to lack of privilege.

**Comment by ph-fritsche** [comment]
@babramczyk 
If the `App` in `failing.test.js` is exactly the component under test in your local setup, I have no explanation why there is an act warning.
If it is only a part of the real component under test, look into what you do in the `useEffect` beside triggering the mock function.
If your `useEffect` causes a rerender, you've got your culprit.

**Comment by anilanar** [comment]
After several years of fighting against `act`, I have the firm belief that `act` is sadly an anti-pattern. It forces tests to know about how rendering happens. Tests need to know whether is there's a `useEffect` that triggers an async rerendering or not. In my book and in many books, that's called an anti-pattern. The community, perhaps over-trusting community leaders, has been unquestionably sprinkling act and waitFor here and there to silence these warnings.

Now that I let that out my system, I can get back to the subject :smiley_cat:  :

It seems some bugs are fixed in version 14.0.0 and I stop getting act warnings after upgrading.
